### DISCUSSION TITLE
Issues with damage increasing Oils

### DISCUSSION THREAD

**Discussion by thebigMuh** [diasurgical/DevilutionX#2022] [Show and tell]
Oil effects were wiped in original Hellfire when starting a new game, but they stick around in DevilutionX (a very welcome change). However, this introduces some long term issues, also in combination with Weird and Gloomy shrines. These are all connected, and at the moment I'm fishing for feedback about which of those issues I should fix (and how).

**Mechanisms right now**
- Oil of Sharpness (OoS): Increases max. damage by 1, as long as the current difference between minimum and maximum damage on the weapon is less than 30.
- Oil of Death (OoD): Increases min. damage by 1 and max. damage by 2, as long as the current difference between minimum and maximum damage on the weapon is less than 30.
- Gloomy Shrine (GS): Reduces maximum damage on all weapons by 1.
- Weird Shrine (WS): Increases maximum damage on all weapons by 1.

**Issues**

***Overflow***
The GS limits the reduction to min. damage, so at worst you will get a weapon that does 1-1 damage (or whatever the minimum is). WS, OoS and OoD don't limit the increase to 255 however, at which point maximum damage will wrap around to zero. This doesn't lead to crashes, it just looks odd. I didn't check how all the calculations behave, but I whacked a few monsters and they took damage.

*Suggestion*
Maybe limit all increases to a weapon's max damage to 255. The chance to actually reach the overflow is slim, but if someone is determined they will eventually achieve it.

***Endless Increase***
Applying OoD will increase the difference between max and min steadily, until the range is 30, and no more OoD can be applied. If a GS is then used, it will reduce the range to 29, which allows applying one more OoD, increasing both minimum and maximum damage. By repeating this, weapon damage can be boosted forever (or until 255 is reached).
Example:
- Bastard Sword, base damage 6-15. 
- 21 OoD can be applied to make it deal 27-57 damage.
- Clicking a GS changes this to 27-56.
- Applying OoD boosts this to 28-58.
- GS -> 28-57
- OoD -> 29-59
- GS -> 29-58
- OoD -> 30-60
Rinse, repeat.

*Suggestion*
This is so stupid that if someone really wants to do it, why not. Wouldn't fix it.

***OoS/WS precludes usage of OoD***
Applying OoS and using WS repeatedly can boost weapon damage until the damage range exceeds 30. At this point, applying OoD to the weapon will not work anymore. This isn't improbable, and it happened to me. OoS are relatively common, so boosting e.g. a Bastard Sword to do 6-36 damage isn't too difficult. Then finding one of the rare OoD and trying to apply it to the weapon without any effect is a big letdown.

*Suggestion*
I would fix this by allowing to use exactly as many OoD on a weapon as if no OoS or WS had been used, essentially making the order of application irrelevant.

**Comment by julealgon** [comment]
> but I whacked a few monsters and they took damage.

Your final damage includes a weaponless damage component, so even if the weapon does net 0 damage, you'll still do (a lot of, if high level) damage. I think this is silly, but it's how the original damage formula works. Check Jarulf's for the actual formula for the intrinsic physical component.

> Maybe limit all increases to a weapon's max damage to 255.

Even if we don't have a clear decision for how it should work in general in regards to limits, values should _never_ overflow, for any value in the entire game. All values should be capped and checked against the cap on increment. Similarly, negative values should _also_ never be allowed. This means every increment and decrement of any stat or resource value should always be bounds checked.

> difference between minimum and maximum damage on the weapon is less than 30

I honestly don't understand the reasoning behind this strategy at all. At best, it seems completely arbitrary to me.

While i have some ideas for a mod (for example, losing the effects of oils incrementally while repairing) for devilutionx I'm not entirely sure what we can do....

Maybe (and this is probably a bit outlandish) the damage limit should be based on the base weapon damage. So a bastard sword, which is normally 6-15, could never surpass 12-30 damage through oils for example (2x base damage). Or even, the max could be the average base damage instead, so a bastard sword could never do more than 21 average damage. 

Then, I'd make the oils randomly add minimum or maximum damage, instead of targeting one or the other with fixed values. For example, oil of sharpness could add +1 damage (randomly selected between min or max) while oil of death would add +3 (+3min, +2min/+1max, +1min/+2max or +3max).

**Reply by thebigMuh** [reply]
> > but I whacked a few monsters and they took damage.
> 
> Your final damage includes a weaponless damage component, so even if the weapon does net 0 damage, you'll still do (a lot of, if high level) damage. I think this is silly, but it's how the original damage formula works. Check Jarulf's for the actual formula for the intrinsic physical component.

Sure, I just wanted to know if the game instantly implodes if max damage < min damage, and it doesn't.

> > Maybe limit all increases to a weapon's max damage to 255.
> 
> Even if we don't have a clear decision for how it should work in general in regards to limits, values should _never_ overflow, for any value in the entire game. All values should be capped and checked against the cap on increment. Similarly, negative values should _also_ never be allowed. This means every increment and decrement of any stat or resource value should always be bounds checked.

Well, theoretically possible... I'll leave it at that.

> > difference between minimum and maximum damage on the weapon is less than 30
> 
> I honestly don't understand the reasoning behind this strategy at all. At best, it seems completely arbitrary to me.

Yes, it's weird and not balanced. I'm sure we can all come up with various better schemes, but for now I mainly want to get rid of the lockout that can happen by first applying OoS, and later trying to upgrade with OoD.

**Comment by NikoVP** [comment]
I'd just fix it so it's consistent. Although I love maxing stuff out I think its stupid going for 256 damage on a weapon. My vote goes for apply the current normal max damage limits and not allow the increase more than intended. At most, maybe allow increasing the min damage after max damage limit has been reached until half the amount of max damage increase in case the latter has been reached by applying OoS or WS.

**Reply by thebigMuh** [reply]
The main thing I want to get rid of is the negative feedback you get right now, when you've upgraded your weapon with OoS, and then can't use OoD. The rest I'll likely leave untouched.

**Reply by NikoVP** [reply]
That's fine too. Unintentional features should be removed

**Comment by julealgon** [comment]
It actually just occurred to me that one could make it so that "different weapons have different potentials for improvement". Say, a bastard sword is normally 6-15, but we could make it so that it has a "maximum potential" of 20-30 or something. At the same time, a dagger is 1-4 but we could have it deal potentially 10-50 or something.

This is probably off-topic here, but it gave me some ideas (the concept of a "max potential damage" for a weapon).

**Reply by julealgon** [reply]
Additionally, "potential damage" could be initially unknown to the player, and become known after use/time like monster stats do. Hmm...

**Reply by thebigMuh** [reply]
The possibilities are limitless. "We have the source code".

That said, I want to do only a fix here :)

**Comment by AJenbo** [comment]
Maybe a reasonable change would be to limit it to the max that could have been rolled by the item generator? Of cause this is a heavy restriction compared to what was possible before, but they also had the bug that would reset it between each game.

**Reply by julealgon** [reply]
> the max that could have been rolled by the item generator?

What exactly do you mean by that, though? Item base damage is never rolled, it's set in stone. Can you provide an example?

**Reply by AJenbo** [reply]
Guess I'm just getting it confused with the powers

**Comment by Psojed** [comment]
You could also implement the bug back as a feature, it works from a realistic standpoint (oil evaporating after a while or smth like that). This solution does make the oils less useful in general, but that can be said about multiple items in the game.

**Comment by FitzRoyX** [comment]
If you could only oil something once for a slight permanent boost ontop of enchantments... that might make sense. But it looks ridiculous to allow oiling things repeatedly. That makes it so that

a) finding oils is more important than finding items themselves
b) you have hidden limits like min+max<30 that require reading a jalruf style guide to understand. who has time for this shit?
c) the limit, once understood, can be exploited
d) you risk wrapping values around

IIRC the game doesn't even show you how much dmg/dur/armor has been gained from oil after it has been applied. Ideally, we show "+1 Damage from Oil" on the attribute window like an enchantment so that you can clearly see an item has been oiled. Yes, I realize you can't do that in the tiny info panel.

**Reply by AJenbo** [reply]
Also it modifies the item so you can't easily extract that info. But it would be nice to show the values. There are some unused flags that we could use to indicate that an item is oiled and the info should be shown.

**Reply by FitzRoyX** [reply]
If it could be done, I'd do it for shrines as well. I don't like the idea of shrines and oils overwriting base values with no way to know which did what and how much.

As far as stacking effects, that's the more arguable change. Oils and shrines can technically be farmed to create super-items that look like something a cheat editor would make. It greatly diminishes the value of found enchantments. And then you have to write code to prevent wrapping bugs.

I would disallow oils of the same type to be stacked on items, and I would cap shrine alterations too. So "+1 from Shrines" is the max blessing and "-1 from Shrines" is the max curse. You're still cursing and blessing items, but you're no longer able to stack the effect into absurdity.

**Reply by AJenbo** [reply]
> If it could be done, I'd do it for shrines as well. I don't like the idea of shrines and oils overwriting base values with no way to know which did what and how much.

You won't know which did what or how much. You will know something happened and what the total is.

> I would disallow oils of the same type to be stacked on items, and I would cap shrine alterations too. So "+1 from Shrines" is the max blessing and "-1 from Shrines" is the max curse. You're still cursing and blessing items, but you're no longer able to stack the effect into absurdity.

Again not really possible since you can really se what the base value is or or what modified the values.

**Comment by ropufu** [comment]
> Oil effects were wiped in original Hellfire when starting a new game, but they stick around in DevilutionX (a very welcome change).

I think some might disagree with the "very welcome" part. What I'd like to see is the ability to switch permanence of *non-durability* oil effects on and off through diablo.ini, with it being disabled by default (faithful to the original Hellfire behavior). Then all of the discussion above is only relevant to players who actually turn it on, and the old-timers get to enjoy the game as it used to be ~~with fewer bugs~~ :^)

**Update:** I phrased poorly. The original where the "with fewer bugs" was not crossed out was misleading.

**Reply by julealgon** [reply]
As always, I'll just mention this:
> ...and the old-timers get to enjoy the game as it used to be...

The boat has sailed on this. The game is already not the same as it used to be for numerous other reasons, so this is a poor argument against the feature to me.

**Reply by AJenbo** [reply]
Items resetting was a limitation stemming from the save game format. The reason durability was permanent as oppose to other oils is that the packed format contains this value (see `PkItemStruct`). From what you are saying it sounds like you think solving that issue is a bug?

Having this as an ini setting isn't simple since it part of how we handle conversion of save games, regenerating items could lead to them morphing.

**Reply by ropufu** [reply]
@julealgon All I'm saying is that it came as a surprise to me. And I did not claim that my argument is a reason to take any action. Just described how I felt---that is all.
@AJenbo Interesting, thank you for the details! No, I do not think it is a bug, not at all. Nor that it should be addressed, especially given how complicated it might be.

**Update:** by "with fewer bugs" I meant that DevilutionX has fixed a lot of bugs compared to Diablo/Hellfier. I phrased poorly. You can cross it from my original message to get a better version of what I was trying to get at.
